File: ruby/034_ThereAreFewTypesOfElements.rb
# 034 - There are few types of elements（★4）
# 尺取法

N, K = gets.split.map(&:to_i)
A = gets.split.map(&:to_i)

use_cnt = Hash.new(0)
ans = cnt = right = 0

N.times do |left|
  while right < N
    a = A[right]
    if use_cnt[a] == 0
      break if cnt == K
      cnt += 1
    end

    use_cnt[a] += 1
    right += 1
  end
  ans = [ans, right - left].max
  use_cnt[A[left]] -= 1
  cnt -= 1 if use_cnt[A[left]] == 0
end

puts ans

File: kotlin/034_ThereAreFewTypesOfElements.kt
/**
 * 034 - There are few types of elements（★4）
 * 尺取法
 */

fun main() {

  val (N, K) = readLine()!!.split(" ").map(String::toInt)
  val A = readLine()!!.split(" ").map(String::toInt)

  var ans = 0
  var cnt = 0
  var useCnt = mutableMapOf<Int, Int>().withDefault { 0 }
  var right = 0
  for (left in 0 until N) {
    while (right < N) {
      val a = A[right]
      if (useCnt.getValue(a) == 0) {
        if (cnt == K) break
        cnt++
      }
      useCnt.put(a, useCnt.getValue(a) + 1)
      right++
    }
    ans = Math.max(ans, right - left)
    useCnt.put(A[left], useCnt.getValue(A[left]) - 1)
    if (useCnt[A[left]] == 0) cnt--
  }

  println(ans)
}
